diff -up qtbase-opensource-src-5.0.1/src/gui/painting/qdrawhelper.cpp.lowmem qtbase-opensource-src-5.0.1/src/gui/painting/qdrawhelper.cpp
--- qtbase-opensource-src-5.0.1/src/gui/painting/qdrawhelper.cpp.lowmem	2013-01-29 13:03:00.000000000 -0600
+++ qtbase-opensource-src-5.0.1/src/gui/painting/qdrawhelper.cpp	2013-01-31 08:21:05.424584177 -0600
@@ -42,7 +42,7 @@
 #if defined(__OPTIMIZE__) && !defined(__INTEL_COMPILER) && defined(__GNUC__) \
     && (__GNUC__ * 100 + __GNUC_MINOR__ >= 404)
 // GCC 4.4 supports #pragma GCC optimize and #pragma GCC target
-#  pragma GCC optimize "O3"
+//#  pragma GCC optimize "O3"
 #  if defined(__i386__) && defined(__SSE2__) && !defined(__SSE2_MATH__)
 #   pragma GCC target "fpmath=sse"
 #  endif
