diff -up qtbase-opensource-src-5.0.2/mkspecs/features/data/cmake/Qt5BasicConfig.cmake.in.1096 qtbase-opensource-src-5.0.2/mkspecs/features/data/cmake/Qt5BasicConfig.cmake.in
--- qtbase-opensource-src-5.0.2/mkspecs/features/data/cmake/Qt5BasicConfig.cmake.in.1096	2013-04-08 19:10:52.000000000 -0500
+++ qtbase-opensource-src-5.0.2/mkspecs/features/data/cmake/Qt5BasicConfig.cmake.in	2013-04-10 07:48:35.667195903 -0500
@@ -83,7 +83,7 @@ if (NOT TARGET Qt5::$${CMAKE_MODULE_NAME
             find_package(Qt5${_module_dep}
                 ${_Qt5$${CMAKE_MODULE_NAME}_DEPENDENCIES_FIND_QUIET}
                 ${_Qt5$${CMAKE_MODULE_NAME}_FIND_DEPENDENCIES_REQUIRED}
-                PATHS \"${_qt5$${CMAKE_MODULE_NAME}_install_prefix}\" NO_DEFAULT_PATH
+                PATHS \"${CMAKE_CURRENT_LIST_DIR}/..\" NO_DEFAULT_PATH
             )
         endif()
 
